commands/dkprune.ml

```diff
@@ -0,0 +1,188 @@
+          Dep.(md_deps.deps)
+      end
+
+let handle_name : Basic.name -> unit = fun name ->
+  let md = Basic.md name in
+  let file = Dep.get_file md in
+  handle_file file
+
+let handle_constraints names =
+  Dep.NameSet.iter handle_name names
+
+let output_file : Dep.path -> Dep.path = fun file ->
+  let basename = Filename.basename file in
+  match !output_directory with
+  | None -> raise @@ NoDirectory
+  | Some dir -> Filename.concat dir basename
+
+let get_files : unit -> (mident * Dep.path * Dep.path) list = fun () ->
+  Hashtbl.fold (fun md _ l ->
+      try
+        let f = Dep.get_file md in
+        (md, f, output_file f)::l
+    with _ -> l) Dep.deps []
+
+let name_of_entry md = function
+  | Entry.Decl(_,id,_,_) ->
+    mk_name md id
+  | Entry.Def(_,id,_,_,_) ->
+    mk_name md id
+  | Entry.Rules(_,r::_) ->
+    let open Rule in
+    let r' = to_rule_infos r in
+    r'.cst
+  | _ as e -> raise @@ EntryNotHandled e
+
+let is_empty deps md in_file =
+  let input = open_in in_file in
+  let empty = ref true in
+  let mk_entry e =
+    let name = name_of_entry md e in
+    if Dep.NameSet.mem name deps then
+      empty := false;
+  in
+  Parser.Parse_channel.handle md mk_entry input;
+  close_in input;
+  !empty
+
+let mk_file deps (md,in_file,out_file) =
+  log "[WRITING FILE] %s" out_file;
+  if not (is_empty deps md in_file)
+  then
+    let input = open_in in_file in
+    let output = open_out out_file in
+    let fmt = Format.formatter_of_out_channel output in
+    let handle_entry e =
+      let name = name_of_entry md e in
+      CustomSig.set md;
+      if Dep.NameSet.mem name deps
+      then Format.fprintf fmt "%a" Printer.print_entry e
+    in
+    Parser.Parse_channel.handle md handle_entry input;
+    close_out output;
+    close_in input
+
+let print_dependencies names =
+  let open Dep in
+  NameSet.iter Dep.transitive_closure names;
+  let down = NameSet.fold
+      (fun name dependencies ->
+         NameSet.union (get_data name).down dependencies) names names in
+  let files = get_files () in
+  List.iter (mk_file down) files
+
+(* This opens a module and returns all the names of symbols declared inside *)
+let names_of_md md =
+  let file = Dep.get_file md in
+  let input = open_in file in
+  let names = ref Dep.NameSet.empty in
+  let mk_entry e =
+    let n = name_of_entry md e in
+    names := Dep.NameSet.add n !names
+  in
+  Parser.Parse_channel.handle md mk_entry input;
+  close_in input;
+  !names
+
+(* This gather all symbols *)
+let gather_names =
+  function
+  | Entry.Require(_,md) -> names_of_md md
+  | Entry.DTree(_,Some md, id) -> Dep.NameSet.singleton (mk_name md id)
+  | _ -> raise BadFormat
+
+let parse_constraints file =
+  let md = mk_mident file in
+  let input = open_in file in
+  let pcstr = List.map gather_names (Parser.Parse_channel.parse md input) in
+  close_in input;
+  List.fold_left Dep.NameSet.union Dep.NameSet.empty pcstr
+
+let _ =
+  let args = Arg.align
+    [ ( "-l"
```

Comment:
`--verbose`?

Reply:
I am not sure about this one. Here I use `-l` as a shortcut to print `dkprune` logging output. While `verbose` would do more like printing logging from `dkprune` but `typing`, `reduction` etc...